### PR TITLE
[WFCORE-4314] - verbose and recursive parameter in read-alias(es)

### PR DESCRIPTION
**Pull request by baranowb**
Issue: https://issues.jboss.org/browse/WFCORE-4314
Related: https://github.com/wildfly/wildfly-core/pull/5008

Improve read commands for elytron keystore.
Resubmit after rebase and few tweaks to doc.

This will require version bump.

### CODE REVIEW

**Comment by fjuma** [comment]
@baranowb Just FYI, I've submitted a PR for the Elytron subsystem bump:

https://github.com/wildfly/wildfly-core/pull/4980

File: elytron/src/main/java/org/wildfly/extension/elytron/CertificateChainAttributeDefinitions.java [hunk 1]
```diff
@@ -115,17 +115,25 @@ static ObjectListAttributeDefinition getNamedCertificateList(final String name)
     }
 
     static void writeCertificate(final ModelNode certificateModel, final Certificate certificate) throws CertificateEncodingException, NoSuchAlgorithmException {
-        certificateModel.get(ElytronDescriptionConstants.TYPE).set(certificate.getType());
+        writeCertificate(certificateModel, certificate, false);
```
Comment (fjuma) [inline elytron/src/main/java/org/wildfly/extension/elytron/CertificateChainAttributeDefinitions.java:118]: Shouldn't this be `true` to keep the current behaviour by default?

File: elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java [hunk 2]
```diff
@@ -101,9 +113,17 @@ protected void executeRuntimeStep(OperationContext context, ModelNode operation)
                 ModelNode result = context.getResult();
                 Enumeration<String> aliases = keyStore.aliases();
                 while (aliases.hasMoreElements()) {
-                    result.add(aliases.nextElement());
+                    final String alias = aliases.nextElement();
+                    final boolean recursive = RECURSIVE.resolveModelAttribute(context, operation).asBoolean();
```
Comment (fjuma) [inline elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java:117]: This variable declaration can be moved outside the `while` loop.

File: elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java [hunk 3]
```diff
@@ -101,9 +113,17 @@ protected void executeRuntimeStep(OperationContext context, ModelNode operation)
...
+                    if(recursive) {
+                        final ModelNode aliasNode = result.get(alias);
+                        final boolean verbose = VERBOSE.resolveModelAttribute(context, operation).asBoolean();
```
Comment (fjuma) [inline elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java:120]: This variable declaration can be moved outside the `while` loop.

File: elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java [hunk 4]
```diff
@@ -85,8 +85,20 @@ public void registerOperations(ManagementResourceRegistration resourceRegistrati
 
     static class ReadAliasesHandler extends ElytronRuntimeOnlyHandler {
 
+        static final SimpleAttributeDefinition RECURSIVE = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.RECURSIVE, ModelType.BOOLEAN, true)
+                .setAllowExpression(false)
```
Comment (fjuma) [inline elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java:89]: This could be `.setAllowExpression(true)`.

File: elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java [hunk 5]
```diff
@@ -85,8 +85,20 @@ public void registerOperations(ManagementResourceRegistration resourceRegistrati
...
+                .setDefaultValue(ModelNode.FALSE)
+                .build();
+
+        static final SimpleAttributeDefinition VERBOSE = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.VERBOSE, ModelType.BOOLEAN, true)
+                .setAllowExpression(false)
```
Comment (fjuma) [inline elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java:94]: This could be `.setAllowExpression(true)`.

File: elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties [hunk 6]
```diff
@@ -1139,7 +1139,10 @@ elytron.key-store.alias.certificate-chain.version=The certificate version.
 
 elytron.modifiable-key-store.read-alias=Read an alias from a KeyStore.
 elytron.modifiable-key-store.read-alias.alias=The alias of the KeyStore item to read.
+elytron.modifiable-key-store.read-alias.verbose=Instruct read to show HEX format fields content.
```
Comment (fjuma) [inline elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties:1142]: Even with `verbose` set to `false`, there are still some HEX formatted fields in the output. So maybe we could reword this to something like:

Whether or not to include the public key and encoded form of a certificate in the output. The default value is true.

File: elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties [hunk 7]
```diff
@@ -1139,7 +1139,10 @@ elytron.key-store.alias.certificate-chain.version=The certificate version.
...
 elytron.modifiable-key-store.read-aliases=Read aliases from a KeyStore.
+elytron.modifiable-key-store.read-aliases.recursive=Read more than name of aliases in keystore.
+elytron.modifiable-key-store.read-aliases.verbose=Include HEX form fields in case of recursive read.
```
Comment (fjuma) [inline elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties:1145]: Even with `verbose` set to `false`, there are still some HEX formatted fields in the output. So maybe we could reword this to something like:

Whether or not to include the public key and encoded form of a certificate in the output. The default value is true.

File: elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties [hunk 8]
```diff
@@ -1139,7 +1139,10 @@ elytron.key-store.alias.certificate-chain.version=The certificate version.
 
 elytron.modifiable-key-store.read-alias=Read an alias from a KeyStore.
 elytron.modifiable-key-store.read-alias.alias=The alias of the KeyStore item to read.
+elytron.modifiable-key-store.read-alias.verbose=Instruct read to show HEX format fields content.
 elytron.modifiable-key-store.read-aliases=Read aliases from a KeyStore.
+elytron.modifiable-key-store.read-aliases.recursive=Read more than name of aliases in keystore.
```
Comment (fjuma) [inline elytron/src/main/resources/org/wildfly/extension/elytron/LocalDescriptions.properties:1144]: Maybe we could reword this to something like:

Include information about each alias in the KeyStore. The default value is false.

File: elytron/src/test/java/org/wildfly/extension/elytron/KeyStoresTestCase.java [hunk 9]
```diff
@@ -430,6 +430,67 @@ public void testKeystoreCli() throws Exception {
...
+        operation.get(ElytronDescriptionConstants.PATH).set(resources + "/firefly-copy.keystore");
+        operation.get(ElytronDescriptionConstants.TYPE).set("JKS");
+        operation.get(CredentialReference.CREDENTIAL_REFERENCE).get(CredentialReference.CLEAR_TEXT).set("Elytron");
+        assertSuccess(services.executeOperation(operation));
+
+        operation = new ModelNode();
+        operation.get(ClientConstants.OP_ADDR).add("subsystem","elytron").add("key-store","ModifiedKeyStore");
+        operation.get(ClientConstants.OP).set(ElytronDescriptionConstants.READ_ALIASES);
+        List<ModelNode> nodes = assertSuccess(services.executeOperation(operation)).get(ClientConstants.RESULT).asList();
+        assertEquals(2, nodes.size());
```
Comment (fjuma) [inline elytron/src/test/java/org/wildfly/extension/elytron/KeyStoresTestCase.java:451]: Would be good to verify the actual contents of the result.

File: elytron/src/test/java/org/wildfly/extension/elytron/KeyStoresTestCase.java [hunk 10]
```diff
@@ -430,6 +430,67 @@ public void testKeystoreCli() throws Exception {
...
+            readOperation.get(ClientConstants.OP_ADDR).add("subsystem","elytron").add("key-store","ModifiedKeyStore");
+            readOperation.get(ClientConstants.OP).set(ElytronDescriptionConstants.READ_ALIAS);
+            readOperation.get(ElytronDescriptionConstants.VERBOSE).set(true);
+            readOperation.get(ElytronDescriptionConstants.ALIAS).set(alias);
+            final ModelNode aliasReadNode = assertSuccess(services.executeOperation(readOperation)).get(ClientConstants.RESULT);
+            assertEquals(aliasReadNode, node);
+        }
+    }
+
+    private void checkCertificaite(final ModelNode certificate) {
```
Comment (fjuma) [inline elytron/src/test/java/org/wildfly/extension/elytron/KeyStoresTestCase.java:485]: `checkCertificate`

File: elytron/src/test/java/org/wildfly/extension/elytron/KeyStoresTestCase.java [hunk 11]
```diff
@@ -430,6 +430,67 @@ public void testKeystoreCli() throws Exception {
...
+        operation = new ModelNode();
+        operation.get(ClientConstants.OP_ADDR).add("subsystem","elytron").add("key-store","ModifiedKeyStore");
+        operation.get(ClientConstants.OP).set(ElytronDescriptionConstants.READ_ALIASES);
+        operation.get(ElytronDescriptionConstants.VERBOSE).set(true);
+        operation.get(ElytronDescriptionConstants.RECURSIVE).set(true);
+        nodes = assertSuccess(services.executeOperation(operation)).get(ClientConstants.RESULT).asList();
+        assertEquals(2, nodes.size());
+
+        for(ModelNode node : nodes) {
+            //assertEquals(node.get(0)+"",1, nodes.size());
```
Comment (fjuma) [inline elytron/src/test/java/org/wildfly/extension/elytron/KeyStoresTestCase.java:462]: Looks like this line can be removed.

File: elytron/src/test/java/org/wildfly/extension/elytron/KeyStoresTestCase.java [hunk 12]
```diff
@@ -430,6 +430,67 @@ public void testKeystoreCli() throws Exception {
...
+                    checkCertificaite(certificateFromChain);
+            } else {
+                //need to clean after above .get
+                node.remove(ElytronDescriptionConstants.CERTIFICATE_CHAIN);
+                checkCertificaite(node.get(ElytronDescriptionConstants.CERTIFICATE));
+            }
+
+            final ModelNode readOperation = new ModelNode();
+            readOperation.get(ClientConstants.OP_ADDR).add("subsystem","elytron").add("key-store","ModifiedKeyStore");
+            readOperation.get(ClientConstants.OP).set(ElytronDescriptionConstants.READ_ALIAS);
```
Comment (fjuma) [inline elytron/src/test/java/org/wildfly/extension/elytron/KeyStoresTestCase.java:477]: Would be good to have a few different tests in this file to ensure the following cases are also covered:

* `readAliases` with `recursive` set to `true` and `verbose` set to `false`
* `readAlias` with `verbose` set to `false`

File: elytron/src/test/java/org/wildfly/extension/elytron/KeyStoresTestCase.java [hunk 13]
```diff
@@ -430,6 +430,67 @@ public void testKeystoreCli() throws Exception {
         assertSuccess(services.executeOperation(operation));
     }
 
+    @Test
+    public void testKeystoreReadVerbose() throws Exception {
+        Path resources = Paths.get(KeyStoresTestCase.class.getResource(".").toURI());
+        Files.copy(resources.resolve("firefly.keystore"), resources.resolve("firefly-copy.keystore"), java.nio.file.StandardCopyOption.REPLACE_EXISTING);
```
Comment (fjuma) [inline elytron/src/test/java/org/wildfly/extension/elytron/KeyStoresTestCase.java:436]: Looks like `removeTestFiles` should be updated to delete `firefly-copy.keystore`.

File: elytron/src/main/java/org/wildfly/extension/elytron/ElytronSubsystemTransformers.java [hunk 14]
```diff
@@ -337,6 +337,14 @@ public String getRejectionLogMessage(Map<String, ModelNode> attributes) {
                 .setDiscard(DiscardAttributeChecker.ALWAYS, ElytronDescriptionConstants.ONLY_LEAF_CERT)
                 .setDiscard(DiscardAttributeChecker.ALWAYS, ElytronDescriptionConstants.SOFT_FAIL)
                 .end();
+        builder.addChildResource(PathElement.pathElement(ElytronDescriptionConstants.KEY_STORE))
```
Comment (fjuma) [inline elytron/src/main/java/org/wildfly/extension/elytron/ElytronSubsystemTransformers.java:340]: This block should be moved to the new `from_15_1` method in this file.

File: elytron/src/main/java/org/wildfly/extension/elytron/ElytronSubsystemTransformers.java [hunk 15]
```diff
@@ -337,6 +337,14 @@ public String getRejectionLogMessage(Map<String, ModelNode> attributes) {
...
+                .addOperationTransformationOverride(ElytronDescriptionConstants.READ_ALIAS)
+                .setDiscard(DiscardAttributeChecker.ALWAYS, ModifiableKeyStoreDecorator.ReadAliasHandler.VERBOSE).end();
```
Comment (fjuma) [inline elytron/src/main/java/org/wildfly/extension/elytron/ElytronSubsystemTransformers.java:342]: You can also add:

`.addRejectCheck(new RejectAttributeChecker.SimpleRejectAttributeChecker(ModelNode.FALSE), ElytronDescriptionConstants.VERBOSE)`

File: elytron/src/main/java/org/wildfly/extension/elytron/ElytronSubsystemTransformers.java [hunk 16]
```diff
@@ -337,6 +337,14 @@ public String getRejectionLogMessage(Map<String, ModelNode> attributes) {
...
+        builder.addChildResource(PathElement.pathElement(ElytronDescriptionConstants.KEY_STORE))
+                .addOperationTransformationOverride(ElytronDescriptionConstants.READ_ALIASES)
+                .setDiscard(DiscardAttributeChecker.ALWAYS, ModifiableKeyStoreDecorator.ReadAliasesHandler.VERBOSE)
+                .setDiscard(DiscardAttributeChecker.ALWAYS, ModifiableKeyStoreDecorator.ReadAliasesHandler.RECURSIVE).end();
```
Comment (fjuma) [inline elytron/src/main/java/org/wildfly/extension/elytron/ElytronSubsystemTransformers.java:346]: You can also add:

`.addRejectCheck(new RejectAttributeChecker.SimpleRejectAttributeChecker(ModelNode.TRUE), ElytronDescriptionConstants.VERBOSE)`
`.addRejectCheck(new RejectAttributeChecker.SimpleRejectAttributeChecker(ModelNode.TRUE), ElytronDescriptionConstants.RECURSIVE)`

**Comment by baranowb** [comment]
@fjuma yep, I ca see, updating. Odd thing is git seemingly hid part of those comments.

**Comment by baranowb** [comment]
@fjuma Updated according to review. Im trying to figure out failures, but it makes no sense ATM.

File: elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java [hunk 17]
```diff
@@ -100,10 +112,22 @@ protected void executeRuntimeStep(OperationContext context, ModelNode operation)
...
                 ModelNode result = context.getResult();
                 Enumeration<String> aliases = keyStore.aliases();
+                ModelNode aliasNode = null;
+                boolean verbose = true;
+                String alias = null;
+                boolean recursive = false;
                 while (aliases.hasMoreElements()) {
-                    result.add(aliases.nextElement());
+                    alias = aliases.nextElement();
+                    recursive = RECURSIVE.resolveModelAttribute(context, operation).asBoolean();
```
Comment (fjuma) [inline elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java:121]: This should be moved outside the `while` loop since the value of `RECURSIVE` doesn't change during the loop.

File: elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java [hunk 18]
```diff
@@ -100,10 +112,22 @@ protected void executeRuntimeStep(OperationContext context, ModelNode operation)
...
+                    if(recursive) {
+                        aliasNode = result.get(alias);
+                        verbose = VERBOSE.resolveModelAttribute(context, operation).asBoolean();
```
Comment (fjuma) [inline elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java:124]: This should also be moved outside the `while` loop.

File: elytron/src/main/java/org/wildfly/extension/elytron/ElytronSubsystemTransformers.java [hunk 19]
```diff
@@ -137,6 +137,20 @@ public void registerTransformers(SubsystemTransformerRegistration registration)
     private static void from15_1(ChainedTransformationDescriptionBuilder chainedBuilder) {
         ResourceTransformationDescriptionBuilder builder = chainedBuilder.createBuilder(ELYTRON_15_1_0, ELYTRON_15_0_0);
 
+        builder.addChildResource(PathElement.pathElement(ElytronDescriptionConstants.KEY_STORE))
+        .addOperationTransformationOverride(ElytronDescriptionConstants.READ_ALIAS)
+        .setDiscard(DiscardAttributeChecker.ALWAYS, ModifiableKeyStoreDecorator.ReadAliasHandler.VERBOSE)
```
Comment (fjuma) [inline elytron/src/main/java/org/wildfly/extension/elytron/ElytronSubsystemTransformers.java:142]: This should be `DiscardAttributeChecker.UNDEFINED`

File: elytron/src/main/java/org/wildfly/extension/elytron/ElytronSubsystemTransformers.java [hunk 20]
```diff
@@ -137,6 +137,20 @@ public void registerTransformers(SubsystemTransformerRegistration registration)
...
+        .addRejectCheck(new RejectAttributeChecker.SimpleRejectAttributeChecker(ModelNode.FALSE), ElytronDescriptionConstants.VERBOSE)
+        .end();
+
+        builder.addChildResource(PathElement.pathElement(ElytronDescriptionConstants.KEY_STORE))
+        .addOperationTransformationOverride(ElytronDescriptionConstants.READ_ALIASES)
+        .setDiscard(DiscardAttributeChecker.ALWAYS, ModifiableKeyStoreDecorator.ReadAliasesHandler.VERBOSE)
+        .setDiscard(DiscardAttributeChecker.ALWAYS, ModifiableKeyStoreDecorator.ReadAliasesHandler.RECURSIVE)
```
Comment (fjuma) [inline elytron/src/main/java/org/wildfly/extension/elytron/ElytronSubsystemTransformers.java:149]: Same for the above two.

**Comment by fjuma** [comment]
@baranowb Thanks for the updates! I think there's just one remaining comment on this line that hasn't been addressed yet:

https://github.com/wildfly/wildfly-core/pull/4951/files#diff-52f390718d0bb48e47afa5fe8b9dd8ac3f4d9e8d1cc50445aab804788e652740R121

That line should be moved outside of the while loop since the value of `recursive` doesn't change during the loop.

**Comment by baranowb** [comment]
Im almost 100% I did change it, weird.

File: elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java (hunk 4 above)
Comment (Skyllarr) [inline elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java:89]: Just minor, but setAllowExpression can be set to `true` ?

File: elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java (hunk 5 above)
Comment (Skyllarr) [inline elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java:94]: same here

File: elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java (hunk 4 above)
Comment (Skyllarr) [reply elytron/src/main/java/org/wildfly/extension/elytron/ModifiableKeyStoreDecorator.java:89]: @baranowb Ignore this comment, @fjuma explained to me that since these are only used in operations there is no need to allow expressions for them

**Review by Skyllarr — APPROVED** [review]

**Comment by Skyllarr** [comment]
@baranowb Just a minor but but the 2 commits should be squashed

**Comment by bstansberry** [comment]
Thanks, @baranowb!
